File: recap/TournamentWins/TournamentWins.cpp
#include<cstdio>
int k, r;
int n, m;
double ans;
int main()
{
  scanf("%d%d", &k, &r);
  n=(1<<k)-1; m=(1<<k)-r;
  double p=1.0;
  for(int k=1;(1<<k)-1<=m;k++)
  {
    int sm=m-(1<<k)+2, em=m-(1<<(k-1))+1;
    int sn=n-(1<<k)+2;
    for(int mm=sm, nn=sn;mm<=em;mm++, nn++) p*=(double)(mm)/(double)(nn);
    ans+=p;
  }
  printf("%.5lf\n", ans);
  return 0;
}

File: recap/TournamentWins/TournamentWins.java
import java.text.DecimalFormat;
import java.util.*;

public class TournamentWins {
    public static void main(String[] args) {
        Scanner s = new Scanner(System.in);
        int k = s.nextInt(), r = s.nextInt();
        double ans = 0.0;
        for (int i = 1; i <= k; i++) {
            double cur = 1.0;
            for (int j = 1; j < (1 << i); j++) {
                cur *= (1 << k) - j - r + 1;
                cur /= (1 << k) - j;
            }
            ans += cur;
        }
        DecimalFormat fmt = new DecimalFormat("0.00000");
        System.out.println(fmt.format(ans));
    }
}
